vivid-server/src/modules/websocket/event.gateway.ts

```diff
@@ -45,4 +62,83 @@ export class EventGateway implements OnGatewayConnection {
       if (res) socket.auth = res;
     } catch (err) {}
   }
+
+  @SubscribeMessage('newGame')
+  newGame(@MessageBody() clientId: string) {
+    const client = this.server.sockets.connected[clientId];
+    if (!client) return;
+
+    // Return if client is already in existing game
+    if (clientRooms[clientId] != null) return;
+
+    // Create unique game ID
+    const roomName = 'e372e47c-3649-44c9-9455-c48f84e3d80d'; // TODO remove this - hardcoded for testing
+    // const roomName: string = uuid();
+    clientRooms[clientId] = roomName;
+
+    // Create GameState and save it in states
+    states[roomName] = createGameState();
+
+    // Join client to room
+    client.join(roomName);
+
+    // Init canvas, context and eventlisteners
+    client.emit('init', 1);
+
+    // Set client player number
+    client.number = 1;
+  }
+
+  @SubscribeMessage('joinGame')
+  joinGame(@MessageBody() body: { clientId: string; roomName: string }) {
+    const client = this.server.sockets.connected[body.clientId];
+    if (!client) return;
+
+    // Return if client is already in existing game
+    if (clientRooms[body.clientId] != null) return;
+
+    if (clientRooms[body.roomName]) clientRooms[body.clientId] = body.roomName;
+
+    // Join client to room
+    client.join(body.roomName);
+
+    // Init canvas, context and eventlisteners
+    client.emit('init', 2);
+
+    // Set client player number
+    client.number = 2;
+
+    // Start game
+    this.startGameInterval(body.roomName);
+  }
+
+  @SubscribeMessage('keydown')
+  handleKeydown(@MessageBody() body: { clientId: string; move: number }) {
+    const client = this.server.sockets.connected[body.clientId];
+    if (!client) return;
+
+    // Find room
+    const roomName = clientRooms[body.clientId];
+    if (!roomName) return;
+
+    // Set player move
+    states[roomName].players[client.number - 1].move = body.move;
+  }
+
+  startGameInterval(roomName: string) {
+    // Get clients in game
+    const clients = this.server.sockets.in(roomName);
+
+    const intervalId = setInterval(() => {
+      const winner: number = gameLoop(states[roomName]);
+
+      if (!winner) {
+        // If no winner, render GameState for both clients
+        clients.emit('drawGame', states[roomName]);
+      } else {
+        clients.emit('gameOver', winner);
+        clearInterval(intervalId);
+      }
+    }, 1000 / 50); // 50 FPS
+  }
```

Comment:
I recommend putting most of the logic inside a pong service and pass the client or server sockets through arguments, or something like it.
that way the events themselves here are pretty clean. (just like controllers. controllers and gateways are bare minimum. the services contain the actual logic)